gbpservice/nfp/configurator/agents/generic_config.py

```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
+                    """
+                    kwargs['fail_count'] = kwargs.get('fail_count') + 1
+                    if kwargs.get('fail_count') >= MAX_FAIL_COUNT:
+                        notification_data = self._prepare_notification_data(
+                                                                        ev,
+                                                                        result)
+                        self._sc.poll_event_done(ev)
+                        self._notification(notification_data)
+                elif result == SUCCESS:
+                    """set fail_count to 0 if it had failed earlier even once
+                    """
+                    kwargs['fail_count'] = 0
+        elif ev.id == const.EVENT_CLEAR_HEALTHMONITOR:
+            """Stop current poll event. event.key is vmid which will stop
+               that particular service vm's health monitor
+            """
+            notification_data = self._prepare_notification_data(ev, result)
+            self._sc.poll_event_done(ev)
+            self._notification(notification_data)
+        else:
+            """For other events, irrespective of result send notification"""
+            notification_data = self._prepare_notification_data(ev, result)
             self._notification(notification_data)
 
+    def _prepare_notification_data(self, ev, result):
+        kwargs = ev.data.get('kwargs')
+        request_info = kwargs['request_info']
+        del kwargs['request_info']
+        context = ev.data.get('context')
+
+        # Retrieve notification and remove it from context. Context is used
+        # as transport from batch processing function to this last event
+        # processing function. To keep the context unchanged, delete the
+        # notification_data before invoking driver API.
+        notification_data = context.get('notification_data')
+        del context['notification_data']
+        resource = context.get('resource')
+        del context['resource']
+
+        msg = {'receiver': const.ORCHESTRATOR,
+               'resource': resource,
+               'method': ev.id,
+               'kwargs': [
+                          {
+                           'context': context,
+                           'resource': resource,
+                           'request_info': request_info,
+                           'result': result
+                          }
+                        ]
+               }
+        if not notification_data:
+            notification_data.update(msg)
+        else:
+            data = {
+                    'context': context,
+                    'resource': resource,
+                    'request_info': request_info,
+                    'result': result
+                   }
+            notification_data['kwargs'].append(data)
+        return notification_data
+
+    def poll_event_cancel(self, ev):
+        LOG.error('Poll event cancelled for event %s ' % (ev.data))
+        result = FAILED
```

Comment:
Why do we need to send failed notification when poll event is canceled? The event can be canceled in positive cases also like clear_healthmonitor() API invocation right?